app/lib/s3_tar_writer.rb

```diff
@@ -0,0 +1,32 @@
+require 'rubygems/package'
+
+class S3TarWriter
+  def initialize(s3_dest_url)
+    @s3_dest_url = s3_dest_url
+  end
+
+  def start_streaming
+    @stdin, _stdout, @wait_thr = Open3.popen2("aws", "s3", "cp", "-", @s3_dest_url)
+    # Set stdin to binary mode, since GzipWriter outputs binary.
+    @stdin.binmode
+    @gz = Zlib::GzipWriter.new(@stdin)
+    @tar = Gem::Package::TarWriter.new(@gz)
+  end
+
+  def add_file_with_data(file_path, data)
+    @tar.add_file_simple(file_path, 0o444, data.length) do |io|
```

Comment:
could you add a comment for the `0o444`

Reply:
Done, also changed to 644 so the file owner can write to it.

Reply:
Actually isn't 644 too open?
Who needs read access?

Reply:
Okay, changed to 600 so only the file owner has access.